<!DOCTYPE html>
<html>

<head>
<link rel=stylesheet href=style.css />
<link rel=icon href=CZI-new-logo.png />
</head>

<body>
    <nav role="navigation">
        <!-- 
        Navigation as updated 8/15/2019.
        To make any changes, here, update Navigation.html,
        then rerun UpdateNavigation.py from the docs directory.
        This will copy to all the html files.
        Do not manually edit the navigation section of any other file.
        -->
        <ul>
            <a class="main-nav" href=index.html><li >Shasta</li></a>
                <a href=QuickStart.html><li>Quick Start</li></a>
                
                <li aria-haspopup="true"> How to
                <ul class="dropdown" aria-label="submenu">
                <a href=Running.html><li>Run an assembly</li></a>
                <a href=Docker.html><li>Run an assembly in Docker</li></a>
                <a href=Performance.html><li>Maximize assembly performance</li></a>
                <a href=BuildingFromSource.html><li>Build the code from source</li></a>
                <a href=InspectingResults.html><li>Explore assembly results</li></a>
                <a href=Contributing.html><li>Contribute to Shasta</li></a>
                <a href=ReportingBugs.html><li>Report problems or ask questions</li></a>
                </ul>
            </li>
            <li aria-haspopup="true"> About
                    <ul class="dropdown" aria-label="submenu">
                            <a href=Motivation.html><li>Motivation</li></a>
                            <a href=CurrentStatus.html><li>Current status</li></a>
                            <a href=SupportedPlatforms.html><li>Supported platforms</li></a>
                            <a href=Acknowledgments.html><li>Acknowledgments</li></a>
                            <a href=ComputationalMethods.html><li>Computational methods</li></a>
                            <a href=CommandLineOptions.html><li>Command line options</li></a>
                            <a href=Compatibility.html><li>Compatibility across releases</li></a>
                            <a href=gpu.html><li>GPU acceleration</li></a>
    
                    </ul>
                </li>
    
        </ul>
    </nav>
    <main>
<h1>Current status</h1>

<h3>Functionality</h3>
<p>
In its current implementation, Shasta uses long reads
to perform a mostly haploid assembly. 
Small heterozygous bubbles in the assembly graph are collapsed,
and only the largest heterozygous bubbles are kept.

<h3>Release</h3>
<p>
<a href='https://github.com/chanzuckerberg/shasta/releases/tag/0.1.0'>
An initial release 0.1.0
</a>
of the Shasta assembler is available.

<h3>Performance and cost</h3>
<p>
The Shasta assembler is functional and is
capable of performing a human assembly at typical coverage
in a few hours on a large machine with an appropriate configuration,
using as input long reads from Oxford Nanopore flowcells.

<p>
For example,
at coverage around 60x, an AWS  
<code>x1.32xlarge</code> instance, with 128 virtual processors and 1952 GB of memory,
completes the assembly in about 5 hours.
The compute cost of such an assembly is around $20 at AWS spot market or reserved prices.

<p>
The Shasta assembler, although not designed or optimized for smaller genomes,
can also assemble bacterial genomes in minutes on a typical laptop
(e. g. 1-2 minutes for an <i>E. coli</i> assembly). 

<h3>Accuracy</h3>
<p>
An extensive error analysis and comparison with other long read assemblers
is available 
<a href='https://www.biorxiv.org/content/10.1101/715722v1'>here</a>.
Overall, Shasta assembly quality
is comparable or better than assembly quality achieved by other,
less performant long read assemblers.


<h3>Assembly parameters</h3>
<p>
Default Shasta assembly parameters are optimized for coverage around 60x, 
and require reconfiguration if using much lower or much higher coverage.


<h3>Output</h3>
<p>
The Shasta assembler creates output in 
<a href=https://en.wikipedia.org/wiki/FASTA>FASTA</a> and 
<a href=https://github.com/GFA-spec/GFA-spec/blob/master/GFA1.md>GFA 1.0</a> formats.
It also includes an http server and a Python API 
that can be used to access and visualize
the data structures created during assembly.
This can be useful for testing or debugging,
or for studying assembler behavior at locations of interest.

<h3>Platforms</h3>
<p>
Basic assembly functionality is provided by the Shasta executable 
which runs on most current 64-bit Linux plus macOS and Windows.
The Shasta executable is statically linked, so it has 
no dependencies and does not require
installation.
It can be used immediately upon download, subject to
setting its execute permission bit. 

<p>
Extended functionality, including a Python API that enables Python scripting
and an http server that can be used to inspect assembly details,
are only available on Ubuntu 16.04 and Ubuntu 18.04.
A script to install the necessary dependencies is
provided.

<p>
Functionality to run Shasta under Docker is also provided.


<h3>Documentation</h3>
<p>
Usable documentation for the Shasta executable is provided in the source tree
at <code>shasta/docs</code>,
and in the install tree at <code>shasta-install/docs</code>.
Documentation for the latest code on GitHub is also 
made available online on 
<a href='https://chanzuckerberg.github.io/shasta'>GitHub Pages</a>.

<p>
The documentation also covers less extensively the functionality
provided by the Shasta http server and by the Shasta Python API.


<h3>Development</h3>
<p>
The Shasta assembler is undergoing active development
and is expected to improve in accuracy, performance,
and functionality.
As an open source project, it is open to external contributions.


</main>
</body>
</html>

